other/maximum_subarray.py

```diff
@@ -1,20 +1,24 @@
-def max_subarray(nums: list[int]) -> int:
-    """
-    Returns the subarray with maximum sum
-    >>> max_subarray([1,2,3,4,-2])
+from collections.abc import Sequence
+
+
+def max_subarray_sum(nums: Sequence[int]) -> int:
+    """Returns the maximum possible sum amongst all non - empty subarrays
+
+    Raises:
+      ValueError: when nums is empty.
+
+    >>> max_subarray_sum([1,2,3,4,-2])
     10
-    >>> max_subarray([-2,1,-3,4,-1,2,1,-5,4])
+    >>> max_subarray_sum([-2,1,-3,4,-1,2,1,-5,4])
     6
     """
+    if not nums:
+        raise ValueError("Input sequence should not be empty")
 
     curr_max = ans = nums[0]
 
-    for i in range(1, len(nums)):
-        if curr_max >= 0:
-            curr_max = curr_max + nums[i]
-        else:
-            curr_max = nums[i]
-
+    for num in nums[1:]:
```

Comment:
`nums[1:]`  creates a copy, so this is O(n) extra memory.  You could use https://docs.python.org/3/library/itertools.html#itertools.islice

Reply:
I have resolved this in the latest set of commits.